pkg/blobstore/cassandra/cassandra_blob_access.go

```diff
@@ -0,0 +1,673 @@
+package cassandra
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"math"
+	"math/rand"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/buildbarn/bb-storage/pkg/blobstore"
+	"github.com/buildbarn/bb-storage/pkg/blobstore/buffer"
+	"github.com/buildbarn/bb-storage/pkg/blobstore/slicing"
+	"github.com/buildbarn/bb-storage/pkg/capabilities"
+	bbdigest "github.com/buildbarn/bb-storage/pkg/digest"
+	"github.com/buildbarn/bb-storage/pkg/util"
+
+	remoteexecution "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	"github.com/gocql/gocql"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	writeConsistency            = gocql.LocalQuorum
+	readConsistency             = gocql.LocalQuorum
+	lastAccessUpdateConsistency = gocql.LocalOne
+
+	simultaneousWorkers = 100
+
+	// The number of workers to use for updating the `last_access` field
+	lastAccessUpdateWorkerCount = 50
+)
+
+var (
+	cassandraBlobStorePrometheusMetrics sync.Once
+
+	segmentWriteHist = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "buildbarn",
+		Subsystem: "storage",
+		Name:      "cassandra_segment_write_duration_in_seconds",
+		Help:      "Amount of time in seconds it takes to write each segment to Cassandra, measured in seconds",
+		Buckets:   util.DecimalExponentialBuckets(-3, 6, 2),
+	})
+
+	segmentReadHist = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "buildbarn",
+		Subsystem: "storage",
+		Name:      "cassandra_segment_read_duration_in_seconds",
+		Help:      "Amount of time in seconds it takes to read each segment from Cassandra",
+		Buckets:   util.DecimalExponentialBuckets(-3, 6, 2),
+	})
+
+	totalWritesHist = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "buildbarn",
+		Subsystem: "storage",
+		Name:      "cassandra_total_write_duration_in_seconds",
```

Comment:
Instead of using "total", please specify to what it is "total". `cassandra_blob_write_duration_seconds`?

Again, to me it's also not clear why this metric is needed. We're essentially measuring the duration of the entire call to `Put()`. This is something that MetricsBlobAccess can already do. MetricsBlobAccess already breaks down the metrics by backend type. If that's not sufficient, we should improve the way we set up MetricsBlobAccess. Not add additional metrics on the Cassandra side to compensate.